src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -559,7 +567,117 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                    {
+                        useAddend = false;
+                        std::swap(embMaskOp1Reg, embMaskOp2Reg);
+                    }
+                    else if (targetReg == embMaskOp3Reg)
+                    {
+                        useAddend = false;
+                        std::swap(embMaskOp1Reg, embMaskOp3Reg);
+                    }
+                    else
+                    {
+                        // Nothing to do here
+                    }
+
+                    switch (intrinEmbMask.id)
+                    {
+                        case NI_Sve_FusedMultiplyAdd:
+                            insEmbMask = useAddend ? INS_sve_fmla : INS_sve_fmad;
+                            break;
+
+                        case NI_Sve_FusedMultiplyAddNegated:
+                            insEmbMask = useAddend ? INS_sve_fnmla : INS_sve_fnmad;
+                            break;
+
+                        case NI_Sve_FusedMultiplySubtract:
+                            insEmbMask = useAddend ? INS_sve_fmls : INS_sve_fmsb;
+                            break;
+
+                        case NI_Sve_FusedMultiplySubtractNegated:
+                            insEmbMask = useAddend ? INS_sve_fnmls : INS_sve_fnmsb;
+                            break;
+
+                        case NI_Sve_MultiplyAdd:
+                            insEmbMask = useAddend ? INS_sve_mla : INS_sve_mad;
+                            break;
+
+                        case NI_Sve_MultiplySubtract:
+                            insEmbMask = useAddend ? INS_sve_mls : INS_sve_msb;
+                            break;
+
+                        default:
+                            unreached();
+                    }
+
+                    if (intrin.op3->IsVectorZero())
+                    {
+                        // If `falseReg` is zero, then move the first operand of `intrinEmbMask` in the
+                        // destination using /Z.
+
+                        assert(targetReg != embMaskOp2Reg);
+                        GetEmitter()->emitIns_R_R_R(INS_sve_movprfx, emitSize, targetReg, maskReg, embMaskOp1Reg, opt);
+                    }
+                    else
+                    {
+                        // Below are the considerations we need to handle:
+                        //
+                        // targetReg == falseReg && targetReg == embMaskOp1Reg
+                        //      fmla    Zd, P/m, Zn, Zm
+                        //
+                        // targetReg == falseReg && targetReg != embMaskOp1Reg
+                        //      movprfx target, P/m, embMaskOp1Reg
+                        //      fmla    target, P/m, embMaskOp2Reg, embMaskOp3Reg
+                        //
+                        // targetReg != falseReg && targetReg == embMaskOp1Reg
+                        //      sel     target, P/m, embMaskOp1Reg, falseReg
+                        //      fmla    target, P/m, embMaskOp2Reg, embMaskOp3Reg
```

Comment:
Can't this one be `movprfx, fmla` since the targetReg is `embMaskOp1Reg`?

That should make it safe to do `movprfx embMaskOp1Reg, P/M, falseReg` (or `movprfx embMaskOp1Reg, P/z` in the case that `falseReg` is known to be `VectorZero`) and then the actual `fmla` which updates the active elements, keeping everything correct

Reply:
> That should make it safe to do movprfx embMaskOp1Reg, P/M, falseReg

Not sure if I follow, but won't that overwrite the `embMaskOp1Reg` of active lanes (on which we want to do an operation) with `falseReg`?

We want something like this:

```asm
not p6, p7
movprfx embMaskOp1Reg p6, falseReg
fmla target, p7, embMaskOp2Reg, embMaskOp3Reg
```

Reply:
Oh right. I keep getting this mixed up... I intuitively expect `movprfx` to support setting the inactive elements so that this behavior all works as expected and so that we can always emit a most optimal 2 instruction sequence.

But in practice it instead only sets the active elements so this doesn't work for cases like this and we require `sel; fmla` instead :/